// electron-publish/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Publisher = exports.getCiTag = exports.HttpPublisher = exports.ProgressCallback = exports.SnapStorePublisher = exports.SpacesPublisher = exports.S3Publisher = exports.KeygenPublisher = exports.GitlabPublisher = exports.GitHubPublisher = exports.BitbucketPublisher = void 0;
var bitbucketPublisher_1 = require("./bitbucketPublisher");
Object.defineProperty(exports, "BitbucketPublisher", { enumerable: true, get: function () { return bitbucketPublisher_1.BitbucketPublisher; } });
var gitHubPublisher_1 = require("./gitHubPublisher");
Object.defineProperty(exports, "GitHubPublisher", { enumerable: true, get: function () { return gitHubPublisher_1.GitHubPublisher; } });
var gitlabPublisher_1 = require("./gitlabPublisher");
Object.defineProperty(exports, "GitlabPublisher", { enumerable: true, get: function () { return gitlabPublisher_1.GitlabPublisher; } });
var keygenPublisher_1 = require("./keygenPublisher");
Object.defineProperty(exports, "KeygenPublisher", { enumerable: true, get: function () { return keygenPublisher_1.KeygenPublisher; } });
var s3Publisher_1 = require("./s3/s3Publisher");
Object.defineProperty(exports, "S3Publisher", { enumerable: true, get: function () { return s3Publisher_1.S3Publisher; } });
var spacesPublisher_1 = require("./s3/spacesPublisher");
Object.defineProperty(exports, "SpacesPublisher", { enumerable: true, get: function () { return spacesPublisher_1.SpacesPublisher; } });
var snapStorePublisher_1 = require("./snapStorePublisher");
Object.defineProperty(exports, "SnapStorePublisher", { enumerable: true, get: function () { return snapStorePublisher_1.SnapStorePublisher; } });
var progress_1 = require("./progress");
Object.defineProperty(exports, "ProgressCallback", { enumerable: true, get: function () { return progress_1.ProgressCallback; } });
var httpPublisher_1 = require("./httpPublisher");
Object.defineProperty(exports, "HttpPublisher", { enumerable: true, get: function () { return httpPublisher_1.HttpPublisher; } });
var publisher_1 = require("./publisher");
Object.defineProperty(exports, "getCiTag", { enumerable: true, get: function () { return publisher_1.getCiTag; } });
Object.defineProperty(exports, "Publisher", { enumerable: true, get: function () { return publisher_1.Publisher; } });
